Add user update

maksym-bobmm opened the pull request:
If address autocomplete is frontend feature https://app.asana.com/0/1159181417647996/1163735716422851

Task https://app.asana.com/0/1159181417647996/1162628642777102
Add field to database
Add some tests
Add avatar uploads
Add gem faker
Add user presenter

maksym-bobmm committed:
Add user update

Add field to database
Add some tests
Add avatar uploads
Add gem faker
Add user presenter

maksym-bobmm committed:
Remove unneeded routes from registration

drkmen commented on app/controllers/application_controller.rb:
```diff
@@ -27,6 +27,7 @@ def render_error_response(message = 'Bad Request', status = :bad_request)
 
   def configure_permitted_parameters
     added_attrs = %i[phone email password password_confirmation]
+    added_attrs2 = %i[phone email password password_confirmation avatar]
```
?
```suggestion
    added_attrs = %i[phone email password password_confirmation avatar]
```

drkmen commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -2,18 +2,30 @@
 
 class Users::RegistrationsController < Devise::RegistrationsController
   skip_before_action :check_authorize, only: [:create]
+  skip_before_action :authenticate_scope!, only: [:update]
```
remove array or use `%i[]`

drkmen commented on app/models/user.rb:
```diff
@@ -21,7 +23,7 @@ def login
   def self.find_for_database_authentication(warden_conditions)
     conditions = warden_conditions.dup
-    if login = conditions.delete(:login)
+    if (login = conditions.delete(:login))
```
why do you add brackets?

maksym-bobmm replied:
RubyMine suggested add clarifying brackets and I did it

drkmen commented on app/services/users/registration/create_service.rb:
```diff
@@ -0,0 +1,19 @@
+      # - password: [String] User password
+
+      def call
+        user = User.new(@params)
```
```suggestion
        user = User.new(params)
```

drkmen commented on app/services/users/registration/update_service.rb:
```diff
@@ -0,0 +1,33 @@
+      def receive_user
+        decoded = JsonWebToken.decode @params['token']
+        User.find(decoded['user_id'])
+      end
```
```suggestion
      def receive_user
        decoded = JsonWebToken.decode params['token']
        @receive_user ||= User.find(decoded['user_id'])
      end
```

drkmen replied:
```suggestion
      def receive_user
        decoded = JsonWebToken.decode params['token']
        @receive_user ||= User.find(decoded['user_id'])
      end
```

drkmen commented on app/services/users/registration/update_service.rb:
```diff
@@ -0,0 +1,33 @@
+      # - token: [String] Token from request
+
+      def call
+        user = receive_user
```
remove this line and use directly `receive_user` method

drkmen commented on spec/controllers/users/registrations_controller_spec.rb:
```diff
@@ -0,0 +1,113 @@
+    it { is_expected.to route(:post, '/api/users/signup').to(action: :create, format: :json) }
+    it { is_expected.to route(:put, '/api/users/signup').to(action: :update, format: :json) }
+    it { is_expected.to route(:patch, '/api/users/signup').to(action: :update, format: :json) }
+
```
remove empty line

drkmen commented on spec/controllers/users/registrations_controller_spec.rb:
```diff
@@ -0,0 +1,113 @@
+    it { is_expected.to route(:put, '/api/users/signup').to(action: :update, format: :json) }
+    it { is_expected.to route(:patch, '/api/users/signup').to(action: :update, format: :json) }
+
+  end
```
add empty line after `end`

drkmen replied:
add empty line after `end`

drkmen commented on spec/controllers/users/registrations_controller_spec.rb:
```diff
@@ -0,0 +1,113 @@
+      end
+      it 'user without email and phone does not change User count' do
+        expect { create_user_without_phone_email }.to_not change(User, :count)
+      end
```
this block looks weird
try to use `context`
```ruby
let(:send_request) { post :create, format: :json, params: params } 
let(:email) { 'email' } 
let(:params) do
 {
  email: email
  ...
 }
end

describe 'POST#create' do
  it 'has 200 code'

  context 'when email is missing' do
    let(email) { nil }

    it 'doesn\'t creates user'
  end
end
```

drkmen commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,23 @@
+  factory :user do
+    password { 'password' }
+
+    factory :user_with_email_phone do
```
use `trait`

drkmen commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,23 @@
+      email { Faker::Internet.email }
+      phone { Faker::Base.numerify('###-###-####') }
+    end
+    factory :user_with_missing_fields_update do
```
same

drkmen commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,23 @@
+      avatar { Rack::Test::UploadedFile.new('/home/developer/avatar.svg') }
+      address { Faker::Address.full_address }
+
+      factory :correct_user_update do
```
and same

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -32,5 +36,31 @@
     end
+
+    context 'user' do
+      context 'on create' do
```
`describe`, `context` and `it` together should forms valid eng sentence
please read: http://www.betterspecs.org/

maksym-bobmm committed:
Add tests for services, controllers, presenter

Add address autocomplete
Remove unneded routes for session

maksym-bobmm committed:
Fix

drkmen commented on spec/controllers/users/registrations_controller_spec.rb:
```diff
@@ -0,0 +1,152 @@
+    let(:create_request) do
+      post :create, params: { user: params[:user].slice(:email, :phone) }, as: :json
+      request.headers['token'] = JSON.parse(response.body)['data']['token']
+    end
```
why can't you just create user using factory_bot?

maksym-bobmm replied:
I did it because I need token for update request, and I can`t get it from creating through factory. But I can get token from create action. In fact that was not needed, because I can get token with JsonWebToken.encode

drkmen replied:
you can also stub it, for example

```ruby
class ApplcationController
  def current_user
    User.find ...
  end
end

#rspec
let(:user) { create(:user) } 
it do
  allow(controller).to receive(:current_user).and_return(user)
  #...
end
```

drkmen replied:
this is still not fixed

drkmen commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,28 @@
+      address_first_last_names
+    end
+  end
+end
```
you need to create a factory, which will build full object.
So, default factory `:user` should contain all required for user fields
traits needed, when you want to redefine something, for example:

```ruby
factory :user do
  name { 'name' }
  active { true }

  trait :inactive do
   # this trait inherits all fields from the parent one
   active { false }
  end
end
```

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -1,35 +1,68 @@
+    context 'on POST#create' do
+      let(:user_with_email) { build(:user, :email) }
+      let(:user_with_phone) { build(:user, :phone) }
+      let(:user_without_email_phone) { build(:user) }
```
regarding factories and traits i described above, here you'll do next:
```ruby
let(:user) { build(:user) }
let(:user_without_email_and_phone) { build(:user, phone: nil, email: nil) }
```

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -1,35 +1,68 @@
+      let(:user_without_email_phone) { build(:user) }
+
+      context 'is successful' do
+        it 'with email' do
```
```
context 'when email is set' do
  it 'valid' do
```

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -1,35 +1,68 @@
-      %w[@example.com example.com user@example@com].each do |email|
-        it { is_expected.to_not allow_value(email).for(:email) }
+      context 'is failed' do
+        it 'without email and phone' do
```
same

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -1,35 +1,68 @@
-    context 'phone' do
-      %w[123-456-7890].each do |phone|
-        it { is_expected.to allow_value(phone).for(:phone) }
+    context 'on POST#update' do
```
what is `POST#update` for user model?

maksym-bobmm replied:
User creates with only email/phone and password. On next step missing fields fills, I do this with update action. There are different validations for create and update actions and I tried test validations for both actions.

drkmen replied:
`POST#update` means a POST request to update action, please rename it. Model is not related to any controller.

drkmen commented on spec/support/shared/registration_controller_shared.rb:
```diff
@@ -0,0 +1,22 @@
+
+RSpec.shared_examples 'with missing fields' do
+  it 'gets 422 code' do
+    create_request
```
please do not do separate request to create something, use factory bot

maksym-bobmm committed:
Fix2

maksym-bobmm committed:
Fix3

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -32,7 +32,7 @@
     end
 
-    context 'on POST#create' do
+    context 'for creating' do
```
when creating

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -51,7 +51,7 @@
       end
     end
-    context 'on POST#update' do
+    context 'for updating' do
```
when updating

drkmen reviewed (changes requested):
presenters and services are must be covered with specs too

maksym-bobmm commented:
Should I rename commits fix/fix2/fix3 and do force push then?

drkmen commented:
>Should I rename commits fix/fix2/fix3 and do force push then?

No, just name commits correctly in the future.

drkmen commented on app/services/users/address_autocomplete_service.rb:
```diff
@@ -0,0 +1,67 @@
+    # more detail https://developers.google.com/places/web-service/autocomplete#place_autocomplete_requests
+
+    def call
+      client = GooglePlacesAutocomplete::Client.new(api_key: ENV['GOOGLE_PLACES_API_KEY'])
```
use separate method for GooglePlacesAutocomplete::Client
```ruby
private

def autocomplete_client
  @autocomplete_client ||= GooglePlacesAutocomplete::Client.new(api_key: ENV['GOOGLE_PLACES_API_KEY'])
end

drkmen commented on spec/controllers/users/address_autocomplete_controller_spec.rb:
```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'support/shared/results_shared'
```
include shared examples folder in `spec/rails_helper.rb`
also rename file to `shared_examples`

drkmen commented on spec/controllers/users/address_autocomplete_controller_spec.rb:
```diff
@@ -0,0 +1,86 @@
+    end
+    let(:input) { 'some search' }
+    let(:google_response) { OpenStruct.new(body: body.to_json) }
+    let(:subject) { post :complete, params: params, as: :json }
```
```ruby
subject { post :complete, params: params, as: :json }
```

drkmen commented on spec/controllers/users/address_autocomplete_controller_spec.rb:
```diff
@@ -0,0 +1,86 @@
+    let(:input) { 'some search' }
+    let(:google_response) { OpenStruct.new(body: body.to_json) }
+    let(:subject) { post :complete, params: params, as: :json }
+    before(:each) do
```
`before`, by default running before each example, so you can to not pass `:each` arg

drkmen commented on spec/controllers/users/sessions_controller_spec.rb:
```diff
@@ -1,7 +1,58 @@
 # frozen_string_literal: true
 
 require 'rails_helper'
+require 'support/shared/results_shared'
```
will no longer needed after you require it in rails_helper

drkmen commented on spec/presenters/user_presenter_spec.rb:
```diff
@@ -0,0 +1,27 @@
+    subject { described_class.new(user).page_context }
+
+    it { is_expected.to be_instance_of Hash }
+  end
```
check what exactly this method returns

maksym-bobmm replied:
![Screenshot at 2020-03-11 11-21-37](https://user-images.githubusercontent.com/52406903/76401484-9c787c80-638a-11ea-9298-d31966996252.png)
hm, it seems it is a hash

drkmen replied:
I mean you need to create an test example which will check returned data.

drkmen commented on spec/services/users/address_autocomplete_service_spec.rb:
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'support/shared/results_shared'
```
same

drkmen commented on spec/services/users/registration/create_service_spec.rb:
```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+require 'support/shared/results_shared'
+
+RSpec.describe 'registration create service' do
```
Users::Registration::CreateService

drkmen commented on spec/services/users/registration/update_service_spec.rb:
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'support/shared/results_shared'
```
same

drkmen commented on spec/services/users/registration/update_service_spec.rb:
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+require 'support/shared/results_shared'
+
+RSpec.describe 'registration update service' do
```
Users::Registration::UpdateService

drkmen commented on spec/services/users/session_service_spec.rb:
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'support/shared/results_shared'
```
same

drkmen commented on spec/services/users/session_service_spec.rb:
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+require 'support/shared/results_shared'
+
+RSpec.describe 'Session service' do
```
Users::SessionService

maksym-bobmm committed:
Requested changes

Some refactoring
Add registration through login
Improve phone regular to prevent pass 123-456-7890sometext

maksym-bobmm committed:
Merge branch 'development' into account_detail-1_2-profile_address

maksym-bobmm committed:
Remove deprecation warning in presenter tests

ActionView::Base instances should be constructed with a lookup context, assignments, and a controller.

maksym-bobmm committed:
Requested fixes

Fix db cleaner error

drkmen commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -1,11 +1,15 @@
+
+    # POST api/users/auth/facebook
+    def facebook
+      result = OmniauthFacebookService.perform(params[:token])
```
Service should receive `Hash`.
Does `params[:token]` contain hash?

drkmen commented on app/services/users/registration/create_service.rb:
```diff
@@ -0,0 +1,30 @@
+          key = params['login'].include?('@') ? 'email' : 'phone'
+          params[key] = params.delete('login')
+        end
+        params
```
```ruby
return params unless params['login'].present?
```
and remove `if` condition

drkmen commented on app/services/users/registration/update_service.rb:
```diff
@@ -0,0 +1,31 @@
+
+      private
+
+      def receive_user
```
why just not `user`?

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -6,33 +6,69 @@
-        it { is_expected.to_not allow_value(email).for(:email) }
+
+      context 'and phone is set' do
+        it 'is valid' do
```
and is valid

drkmen commented on spec/models/user_spec.rb:
```diff
@@ -6,33 +6,69 @@
-        it { is_expected.to allow_value(phone).for(:phone) }
+
+    context 'when updating' do
+
```
remove empty line

drkmen commented on spec/services/users/address_autocomplete_service_spec.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+
+    context 'when input' do
+
```
remove empty line

drkmen commented on spec/services/users/registration/create_service_spec.rb:
```diff
@@ -0,0 +1,39 @@
+
+RSpec.describe Users::Registration::CreateService do
+  describe '#call' do
+    subject { Users::Registration::CreateService.new(user_params).call }
```
`described_class.new(user_params).call`

drkmen commented on spec/services/users/registration/update_service_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  describe '#call' do
+    let(:user) { create(:user, :create, phone: nil) }
+    let(:service) { Users::Registration::UpdateService.new(user_params) }
+    subject { service.call }
```
please use same code style as you used in previous service

drkmen commented on spec/services/users/registration/update_service_spec.rb:
```diff
@@ -0,0 +1,47 @@
+
+      it { is_expected.to be_instance_of OpenStruct }
+      it 'changes user' do
+        last_user = User.last
```
`user` variable supposed to be update. I guess you don't need `last_user`

drkmen commented on spec/services/users/registration/update_service_spec.rb:
```diff
@@ -0,0 +1,47 @@
+
+      it { is_expected.to be_instance_of OpenStruct }
+      it 'doesn\'t change user' do
+        last_user = User.last
```
same here

drkmen commented on spec/services/users/session_service_spec.rb:
```diff
@@ -0,0 +1,25 @@
+RSpec.describe Users::SessionService do
+  describe '#call' do
+    context 'when authenticated user' do
+
```
remove empty line

drkmen commented on spec/services/users/session_service_spec.rb:
```diff
@@ -0,0 +1,25 @@
+
+      context 'is provided' do
+        let(:user) { build(:user, :create, email: 'user@example.com', phone: nil) }
+        subject { Users::SessionService.new(user).call }
```
```ruby
subject { described_class.new(user).call }
```

drkmen commented on spec/support/shared/shared_examples.rb:
```diff
@@ -0,0 +1,54 @@
+      .and not_change(user, :last_name)
+      .and not_change(user, :address)
+  end
+end
```
it is really needed? it doesn't look like a shared example

maksym-bobmm replied:
it calls few times from registrations_controller_spec.rb
![image](https://user-images.githubusercontent.com/52406903/76776129-05a72800-67af-11ea-9a25-192af7e84bf0.png)

drkmen replied:
so, then please it directly in `registrations_controller_spec.rb`
that example is very specific and logically not belongs to shared_examples file
